README
Blitzed Services
----------------

$Id$

Blitzed Services are copyright (c) 2000-2002 Blitzed Services Team.  There
is absolutely NO WARRANTY provided with this program.  Blitzed Services may
be freely redistributed; see the GNU General Public License (in the
file "COPYING") for details.

TABLE OF CONTENTS
      1. Credits
      2. Introduction
      3. Reaching The Authors
      4. Public Mailing List
      5. Changes
     

1. CREDITS

     Blitzed Services are based on ircservices 4.4.8 copyright Andy
Church and Andrew Kempe.  Additional development of Blitzed Services has
been done by the Blitzed Services Team for the benefit of the Blitzed IRC
Network.

2. INTRODUCTION

     Please thoroughly read all the ircservices documentation in the
old-docs/ directory, ircservices provide the real backbone of Blitzed and
Blitzed Services as they are now could never have existed without the work that
Andrew Church et al have done on ircservices.  Please do not expect to get
support for Blitzed Services from ircservices mailing lists and developers,
however.  If you need some help you can find us in #blitzed on the server
irc.blitzed.org.

     Blitzed Services are tested and developed solely for dalnet's Bahamut
ircd.  Please don't ask for support of any other ircd.  Provided you have
a network of Bahamut ircds, Blitzed will happily provide nickname and
channel services for you.

3. REACHING THE AUTHORS

We can be found on irc.blitzed.org in #blitzed, though email is much better
for detailed discussion.  You can email us at services@lists.blitzed.org.

4. SERVICES PUBLIC MAILING LIST

     There is also a public mailing list, services@lists.blitzed.org, for 
discussion of, and announcements regarding, Services. The list is unmoderated, 
but you must be subscribed to the list in order to post to it.  To subscribe, 
send an E-mail message to:

	services-request@lists.blitzed.org

with the following in the body of the message:

	subscribe <your-address>

where <your-address> is your E-mail address (without the angle brackets).
Unsubscription works the same way, but use "unsubscribe" instead of
"subscribe".

5. CHANGES

     Here follows a brief list of features/changes.

- Version 3.0.18, 10th September 2002

	Fixed MemoServ crash when MemoServ SEND was used with no arguments.

	Again fixed the "This nickname is itself a linked nick" nick typo, as
	the last change was mistakenly backed out.

	Fixed ChanServ crash when trying to delete an unregistered nick
	from a channel access list.

- Version 3.0.17, 6th September 2002

	Fixed a typo that caused services to crash when NickServ LIST was
	used to view SUSPENDED nicks.

	Fixed a typo that caused garbled "This nickname is itself a linked
	nick" messages.

	Fixed a coding error that would allow anyone to set the last quit
	time and message for any of your linked nicks, as long as you were
	identified to one of the nicks at the time.

	Fixed typo that was preventing AKILLs from expiring properly.

	Added a simple check to stop people trying to run Services with
	MySQL 3.x, which is not supported.

	NSExtraGrace and CSExtraGrace options, which allow long-time users
	more leeway on nick and channel expiry, were implemented.

Massive changes made between November 2001 and August 2002 to produce a version
of ircservices with a custom MySQL backend.  Blitzed Services 3.x is born.

- Version 2.0.14, 29th November 2001

	A Services Admin command to alter the registration time for
	arbitrary nicks was added.

	NickServ REGAIN security problem fixed by recording the unique
	ID of the user issuing the REGAIN command.

	Problems with the display format of ChanServ AKICK VIEW fixed.

	Crash related to channels setting AOP level negative was fixed.

- Version 2.0.13, 16th September 2001

	Services now automatically restarts when it loses contact with its
	uplink.

	All AKILLs are now placed in lowercase, since parts of services and
	ircd were case sensitive.

	Services is now capable of NICKIP, a bahamut extension which
	propogates every user's IP address across the network.  The
	practical benefit of this is that Services can now operate with
	both host names and IP addresses in AKICKs, NickServ ACCESS lists,
	EXCEPTION lists, AKILLS and so on.

	An example crontab script to periodically check and restart
	services has been added.

	ChanServ SYNC was added in order to allow users to more easily
	remove the effects of any desync that may occur.

	The NickServ REGAIN command was disabled due to reported flaws
	that resulted in a nickname being lost.  The command will remain
	disabled until we can get to the bottom of the matter.

	A channel is now regarded as "last used" whenver any member of the
	channel has cause to use their access in any manner.  Previously,
	a channel's "last used" time was only updated when a user became
	opped on join by ChanServ.  This led to expiry of the channel
	after 30 days. (Reported by #thefalls and RuneGod, sorry!)

	Extra ChanServ ACCESS information is now available using ChanServ
	ACCESS VIEW.  Extra information:
		- Time the access was added
		- Who added it (their nick at the time)
		- Time the access was last modified
		- Who modified it (their nick at the time)
		- Time the access was last used
		- Time the owner of the access last read channel memos

	Minor code cleanups for the vop/aop/sop functions and access list
	sorting routines.

	Extra ChanServ AKICK information is now available using ChanServ
	AKICK VIEW:
		- Time AKICK was added
		- Time AKICK was last used

	Better usage notices for NickServ RECOVER and ChanServ OP.
	(Reported by blitz and dg respectively)

	If a channel has new memos since the last time they were read this
	is now reported to users on-join.  A notice is also sent to users
	who are in the channel at the time a memo is sent to the channel.

	The MEMO ChanServ LEVEL setting has been broken up into MEMO-SEND
	and MEMO-READ, so that founders may allow people to read, but not
	send channel memos.

	A new ChanServ LEVEL setting was added: AKICK-LIST.  This allows
	use of the ChanServ AKICK LIST and VIEW commands alone.

	AKILL policy has been changed.  Now any global oper may add an
	AKILL that lasts no more than 12 hours and affects no more than 20
	people at once.  They may remove AKILLs placed by themselves only.
	Services Opers and Services Admins may place AKILLs of arbitrary
	duration affecting any number of users.  Services Admins may
	additionally override the "wildcard threshold" by using the FORCE
	parameter.

- Version 2.0.12, 23rd May 2001

	Removed ability for non-opers to add or remove mode O (oper-only
	channel) in a channel's MLOCK.  Reported by silencer@blitzed.org.

	Updated informational message from NickServ REGISTER which was pointing
	users at a nonexistant help file regarding hiding their email address.

	Fixed typo in original ircservices code which allowed FORBIDDEN
	channels to still be dropped by their founder.

	ChanServ OP command is now more stringent about who it ops when
	channel is set SECURE or SECUREOPS.  If a channel is set SECURE
	then the target nick must be identified to NickServ.  If the
	channel is set SECUREOPS then the target nick needs to be on the
	access list at level AUTOOP or higher.

- Version 2.0.10, 26th February 2001

	Removing code for other ircds introduced two major bugs: ChanServ
	started joining channels in order to enforce AutoKicks, and AutoKills
	no longer expired correctly.  Both these bugs were fixed.

	Minor fixes to help files.

- Version 2.0.9, 15th February 2001

	Users may now change nick between their linked nicknames without
	needing to re-identify to services.

	Fixed problem where AutoKills placed due to exceeding session
	limits (cloning) could not be removed.

	Renamed ChanServ SET OPNOTICE option to VERBOSE.  All uses of
	ChanServ on VERBOSE channels are logged to the channel operators
	in the form of a notice.

	Added support for the new Blitzed channel mode +a (Anonymous Ops).

	Minor changes to helpfiles and MOTD information.

- Version 2.0.7, 31st January 2001

	Andrew Church released ircservices-4.5pre.  Blitzed Services is
	based on ircservices 4.4.8 and so this upgrade mainly merges
	in any applicable or desired new features.

	Support for ircd's other than Bahamut (and blitzed versions of it)
	removed.

	RFC compliant string matching (from ircservices-4.5pre) so that
	nicks and channels with []{}\| in them behave correctly.

	Fixed bug where a channel founder whose nick was dropped or
	expired whilst they were also on the channel's access list would
	leave the access list corrupted. (from ircservices-4.5pre).

	Number of channels registered is now checked on nick link.
	(from ircservices-4.5pre).

	Fixed bug where a channel's successor could also be made the
	founder. (from ircservices-4.5pre).

	Fixed cosmetic bug where setting only +l or +k as a modelock
	would report the modelock as cleared. (from ircservices-4.5pre).

	Fixed bug where ChanServ CLEAR MODES wasn't clearing mode +O

	OperServ STATS ALL now shows separate memory usage for MemoServ.
	(from ircservices-4.5pre).

	Services will now reopen logfiles when it receives SIGUSR2
	(from ircservices-4.5pre).

	Mixed time intervals are now allowed (e.g. 1h30m) (from
	ircservices-4.5pre).

	Simpler Guest-nick generation algorithm. (from ircservices-4.5pre).

	Cleaned up problems with nick linking (from ircservices-4.5pre).

	Fixed "Currently online" being shown in NickServ INFO even when the
	(from ircservices-4.5pre).

	Support for services stamps to enhance split recovery.

	Added support for +a user mode (services admin).  Non-services
	admins trying to set this mode will have it taken from them.  This
	still needs some work, but it basically is working okay.

	Better Turkish translation, courtesy of SuPeRGiRL.
	
- Version 2.0.6, 25th January 2001

	Emergency fixes to NickServ LISTCHANS internals which were causing
	random segmentation faults of services.

- Version 2.0.5, 24th January 2001

	NickServ LISTCHANS command added, which allows users to list all
	channels in which they have access.  Services admins can specify
	any nick.

	NSRegExtraInfo config option added.  An optional string to send
	to users as a notice when they register a nickname.

	NickServ LISTLINKS modified to allow users to use this command
	on their own nicks.

	MemoServ GLOBALSEND, OPERSEND and CSOPSEND added, to send memos
	to all registered nicks, all registered IRC operators and all
	services operators, respectively.

	"ALL" option added to MemoServ READ to allow all memos to be
	sent, whether previously unread or not.

	NickServ LIST now has an option to match nicks set as IRC
	operators.

	OperServ AKILL VIEW now shows the full time and date of the
	AutoKill as opposed to just the date.

- Version 2.0.4, 20th January 2001

	Fixed denial of service when massive ranges were specified, e.g.
	/msg memoserv del 1-9999999999999999 would cause services to
	loop until it pinged out.

	Added a wallops to notify opers of a new max user count.

	Wallops sent when autokills are added and deleted were made more
	verbose.

	NickServ REGAIN command added.  "/msg nickserv regain nick pass"
	will kill "nick", change your nick to nick and then identify you
	to services, provided that your password is correct.

- Version 2.0.2, 9th January 2001

	Fixes to help files after NickServ KILL was removed.

	Fixed wrongly formatted reports to the snoop channel when services
	admins used NickServ SET on other users.

	The appropriate users are now notified immediately when a new item
	of oper news or logon news is added.

	Modifying an access list with the VOP/AOP/SOP commands could cause
	it to become disordered, this has now been fixed.

	As of version 2.0.1 channel access lists were sorted by access
	level, but entries of equal access level were randomly placed.
	Entries with the same level are now sorted alphabetically.

	/info services.* now displays the correct information.

	New channel LEVELS LEV-CHANGE and LEV-LIST allow channel founders
	to decide who can list their levels settings (/msg ChanServ levels
	#channel list) and who can change them.

	ChanServ LIST is now case insensitive.

	ChanServ CLEAR MODES now clears mode +R (only allow registered
	nicknames into the channel).

	NickServ NOEXPIRE has been made to properly log, snoop and wallops
	on successful use.

	A wallops is now sent when SENDPASS is used.

- Version 2.0.1, 3rd January 2001

	VOICE and DEVOICE commands added to ChanServ.

	Channel access lists are now kept sorted by access level.

	VOP/AOP/SOP ADD commands added to ChanServ as shortcuts to the
	usual ACCESS ADD syntax.

	Configuration options added to show a wallops when nicks and
	channels are registered

	Chanserv OP/DEOP commands modified so that if the "nick" parameter
	is missed off they op/deop the person using them.

	NickServ FORBID made to log in snoop channel.

	ChanServ FORBID made to log in snoop channel and to send a
	wallops.

	NickServ SENDPASS made to send a wallops.

	NickServ ACCESS lists can be manipulated by number.

- Version 2.0.0, 29th December 2000

	import-wrecked-db made from existing import-db, correctly imports
	a Wrecked v1.1.6 services database to ircservices format.  New -t
	option sets all nick and channel passwords to "testnet" which is
	useful for allowing developers to play with a real network's
	database.

	NickServ messages regarding nick enforcement were made a bit more
	readable.

	All options regarding killing people who use your nick were
	removed.  Forcing a nickname change was made the default.
	NickServ SET KILL became NickServ SET ENFORCE.

	Services could be forced to segfault when MemoServ commands were
	used on forbidden channels.  This problem was reported by Scott
	Seufert <scotts@ure.net>, additional fixes from Sean Kelly
	<smkelly@zombie.org>.

	Wallops are now sent when services admins:
		- Change user's NickServ settings.
		- Unlink people's nicks.
		- FORBID nicks.
		- DROP nicks.

	SENDPASS was added to NickServ and ChanServ.  This command allows
	a nick or channel password to be emailed to the owner without the
	operator ever seeing the password.  This also required extra config
	options.

	NickServ SET MARK was added.  A MARKED nick may not be GET, SET or
	SENDPASSED except by a services admin.

	NickServ SET IRCOP was added.  Any opered client without this nick
	option set is nagged about it.

	A problem preventing services from being restarted more than once
	under Linux was fixed.

	NickServ SET NOOP was added.  This option prevents people from
	adding you to any channel access list.

	Configuration options were added to send a URL to a user when they
	register a nick or channel, the idea being to tell them on that
	page all of the things they can do with their new nick/channel.

	A dynamic MOTD for services showing some global settings and
	listing available services admins was added.  Visible with the
	/motd services.* command.

	The development libraries of Red Hat 7 produce a services binary
	which segfaults on empty config lines.  This proved to have a
	trivial workaround.

	ircservices does not allow services admins to be AutoKicked.  An
	option which disables this feature was added.

	ircservices gives channel operator status to all services admins in
	every channel.  An option which disables this feature was added.

	An option was added to make the provision of an email address
	mandatory when a nickname is registered.

	Users are optionally informed of the time remaining before the next
	database update, when they register a nick or channel.

	ID and REG shortcuts for IDENTIFY and REGISTER added to both
	NickServ and ChanServ.

	Services clients made to respond to CTCP VERSION.

	Support for an optional "snoop" channel added.  This channel
	receives logging information from the services clients.

	ChanServ REGISTER, IDENTIFY, DROP, AND GETPASS now log to snoop
	channel also.

	NickServ REGISTER, IDENTIFY, DROP, LINK, UNLINK, RECOVER, RELEASE,
	GHOST AND GETPASS now log to snoop channel also.

	All use of OperServ is logged in snoop channel.

	All services clients now have a header and footer on their HELP
	output.

	NickServ INFO for a nick now displays how long ago a nick was
	registered, e.g.:

		"Time registered: Oct 31 05:41:39 2000 GMT (9 weeks, 1 day, 06:42:17 ago)"

	NickServ INFO output given a general overhaul.

	Fixed a few typos.

	Nicks can now be JUPED.

	When a server is JUPED it is SQUIT first.



ChangeLog
2002-09-10 07:48  andy

	* src/channels.c:
	
	Remove debug message.

2002-09-10 01:10  andy

	* configure, configure.in:
	
	Up version number

2002-09-10 01:07  andy

	* src/chanserv.c:
	
	No need to free nickinfo when there isn't a nickinfo!

2002-09-10 01:06  andy

	* src/channels.c:
	
	Temporary debugging log message.

2002-09-08 13:17  andy

	* src/nickserv.c:
	
	Fix again the "nick has X links" message whose fix got backed out
	for some reason..

2002-09-08 13:17  andy

	* src/memoserv.c:
	
	Dumb case of using something without checking if it is null first.

2002-09-06 16:21  andy

	* ChangeLog, README:
	
	Update changelog & README for new release

2002-09-06 15:55  andy

	* configure, configure.in, data/example.conf, src/chanserv.c,
	src/config.c, src/extern.h, src/nickserv.c:
	
	Autoconf update for version increase.
	
	NSExtraGrace and CSExtraGrace implemented as discussed:
	http://lists.blitzed.org/pipermail/services/2002-August/000468.html

2002-09-06 13:26  andy

	* src/mysql.c:
	
	Quick check to stop people running services with mysql 3.x.  Really
	something needs to go in configure.in.

2002-09-06 11:54  andy

	* src/akill.c:
	
	Fix typo - need to check for @ character in row[1] not row[0]! 
	This will fix bug where akills seem to never be expired.

2002-09-02 03:08  andy

	* src/users.c:
	
	Must only update last_seen/last_quit stuff if the *real_id* is
	identified, not the nick_id.  Spotted by Silencer.

2002-09-02 00:37  andy

	* src/nickserv.c:
	
	Fix broken "nick is link" message.

2002-08-14 07:50  andy

	* src/nickserv.c:
	
	Typo.

2002-08-09 16:19  andy

	* configure, configure.in:
	
	up version

2002-08-08 15:30  andy

	* src/: channels.c, chanserv.c, extern.h, services.h:
	
	autolimit and clearbans code rewritten to use only online channel
	data.  autolimit and clearbans data now stored in Channel
	structure, updated on channel creation, destruction, or when
	settings are changed.

2002-08-06 21:33  andy

	* configure, configure.in:
	
	Up version

2002-08-06 19:00  andy

	* src/: extern.h, nickserv.c, users.c:
	
	mysql_update_last_seen() is a better name for
	mysql_identify_nick(), especially because now it needs to set
	different nick status, not just NS_IDENTIFIED.

2002-08-06 18:49  andy

	* src/: extern.h, nickserv.c, users.c, lang/en_us.l, lang/index:
	
	Don't let people link a nick that already has links, to another
	nick.
	
	mysql_identify_nick() function added to centralise code that
	updates the "last seen" columns (last_seen, last_usermask, etc.).
	
	Update "last seen" stuff when someone changes nick to one of their
	linked nicks.

2002-08-06 18:45  andy

	* src/chanserv.c:
	
	Tracked down the source of spurious "out of sync" errors: parts of
	chanserv access list/view do row by row queries and send notices
	inside the while loop.	notice_lang() will do an SQL query if the
	nick cache is out of date or not present for that nick.  To fix
	this, the cache needs to be updated for that nick before hand.

2002-08-01 12:32  andy

	* configure, configure.in, src/mysql.c:
	
	Temporarily turn SQL logging on all the time to try and find a bug.
	
	Up version.

2002-07-30 04:53  andy

	* src/akill.c:
	
	Forgot to send akill when people match it!

2002-07-30 00:49  andy

	* src/chanserv.c:
	
	typo, spotted by ron.

2002-07-28 16:58  phil

	* src/lang/: en_us.l, es.l, it.l, ja_euc.l, ja_jis.l, ja_sjis.l,
	pt.l, tr.l:
	
	added autojoin to the generic list of commands in nickserv help

2002-07-28 00:27  andy

	* configure, configure.in, src/chanserv.c:
	
	Allow non-id'd nicks to id for channels again.	(OK you convinced
	me Andryan)

2002-07-27 10:36  andy

	* src/chanserv.c:
	
	Make sure people are id'd to a nick before they can id to a channel

2002-07-27 10:35  andy

	* src/process.c:
	
	typo

2002-07-25 01:43  andy

	* mysql/blitzed.sql:
	
	New database schema.

2002-07-25 01:41  andy

	* configure, configure.in, data/example.conf, src/config.c,
	src/extern.h, src/init.c, src/memoserv.c, src/misc.c:
	
	rot13 memo content if MSROT13 option is given in config file.  Also
	found and fixed SQL syntax error when there is 1 unread memo and
	check_memo() is called.

2002-07-24 04:10  andy

	* configure, configure.in:
	
	up version

2002-07-24 04:09  andy

	* src/memoserv.c:
	
	Fix memos so they always go to the master nick.

2002-07-23 16:09  andy

	* configure, configure.in, src/akill.c:
	
	Fix akill reasons not being used.

2002-07-23 05:31  andy

	* src/: process.c, sockutil.c:
	
	Added hack to get around the odd -1-in-net-buffer problem when
	rotating logs.

2002-07-23 05:30  andy

	* src/lang/en_us.l:
	
	Add missing LOGSQL help.

2002-07-23 00:59  phil

	* src/lang/en_us.l:
	
	Fixed CHANSERV and NICKSERV password retrieval e-mails to use
	PASSWORD rather than PASS in the example SET command.

2002-07-22 21:49  andy

	* configure, configure.in, src/chanserv.c:
	
	missing return causes crash when auth code is wrong in chanserv
	sendpass.  up version number

2002-07-21 09:37  andy

	* configure, configure.in, src/Makefile.am, src/Makefile.in,
	src/extern.h, src/init.c, src/messages.c, src/mysql.c,
	src/operserv.c, src/version.h, src/version.sh, src/lang/en_us.l,
	src/lang/index:
	
	Add ability to turn mysql logging on and off.
	
	Remove version.sh as we get version stuff from autoconf now. 
	version.h can be made into a static header.

2002-07-21 07:52  andy

	* src/: extern.h, messages.c, operserv.c:
	
	Remove list of "people available to retrieve passwords" from the
	motd.  It was never completely correct and is now useless since
	anyone can sendpass.

2002-07-20 04:01  andy

	* src/: chanserv.c, extern.h, init.c, main.c, memoserv.c,
	messages.c, mysql.c, news.c, nickserv.c, operserv.c, rotatelogs.sh,
	sessions.c, lang/en_us.l:
	
	Added logging of SQL queries, logfile to be cycled when SIGUSR2 is
	caught.
	
	Removed all references to "read only" mode as it isn't used
	anymore.

2002-07-19 04:14  andy

	* src/: rotatelogs.sh, svschk.sh:
	

src/sha1.c
/*
 * SHA1 hash functions.
 *
 * This came from GnuPG:
 *
 * Copyright (C) 1998, 1999, 2000, 2001 Free Software Foundation, Inc.
 *
 * This program is free but copyrighted software; see the file COPYING for
 * details.
 */

#include "services.h"

RCSID("$Id$");

#if defined(__GNUC__) && defined(__i386__)
static inline uint32 rol(uint32 x, int n);
#endif
static void burn_stack(int bytes);
static void transform(SHA1_CONTEXT *hd, const unsigned char *data);
static void sha1_write(SHA1_CONTEXT *hd, const unsigned char *inbuf,
    size_t inlen);
static void sha1_final(SHA1_CONTEXT *hd);
static unsigned char *sha1_read(SHA1_CONTEXT *hd);


/*
 * Rotate a 32 bit integer by n bytes.
 */
#if defined(__GNUC__) && defined(__i386__)
static inline uint32 rol(uint32 x, int n)
{
	__asm__("roll %%cl,%0"
		:"=r" (x)
		:"0" (x),"c" (n));
	return(x);
}
#else
#	define rol(x,n) ( ((x) << (n)) | ((x) >> (32-(n))) )
#endif

static void burn_stack(int bytes)
{
	char buf[128];
    
	memset(buf, 0, sizeof(buf));
	bytes -= sizeof(buf);
	
	if (bytes > 0)
		burn_stack(bytes);
}


void sha1_init(SHA1_CONTEXT *hd)
{
	hd->h0 = 0x67452301;
	hd->h1 = 0xefcdab89;
	hd->h2 = 0x98badcfe;
	hd->h3 = 0x10325476;
	hd->h4 = 0xc3d2e1f0;
	hd->nblocks = 0;
	hd->count = 0;
}


/*
 * Transform the message X which consists of 16 32-bit-words.
 */
static void transform(SHA1_CONTEXT *hd, const unsigned char *data)
{
	uint32 a, b, c, d, e, tm;
	uint32 x[16];

	/* get values from the chaining vars */
	a = hd->h0;
	b = hd->h1;
	c = hd->h2;
	d = hd->h3;
	e = hd->h4;

#ifdef WORDS_BIGENDIAN
	memcpy(x, data, 64);
#else
	{
		int i;
		unsigned char *p2;
		
		for (i = 0, p2 = (unsigned char *)x; i < 16; i++, p2 += 4) {
			p2[3] = *data++;
			p2[2] = *data++;
			p2[1] = *data++;
			p2[0] = *data++;
		}
	}
#endif


#define K1 0x5A827999L
#define K2 0x6ED9EBA1L
#define K3 0x8F1BBCDCL
#define K4 0xCA62C1D6L
#define F1(x,y,z) ( z ^ ( x & ( y ^ z ) ) )
#define F2(x,y,z) ( x ^ y ^ z )
#define F3(x,y,z) ( ( x & y ) | ( z & ( x | y ) ) )
#define F4(x,y,z) ( x ^ y ^ z )


#define M(i) ( tm = x[i&0x0f] ^ x[(i-14)&0x0f] \
    ^ x[(i-8)&0x0f] ^ x[(i-3)&0x0f] \
    , (x[i&0x0f] = rol(tm,1)) )

#define R(a,b,c,d,e,f,k,m)  do { e += rol( a, 5 )     \
				      + f( b, c, d )  \
				      + k	      \
				      + m;	      \
				 b = rol( b, 30 );    \
			       } while(0)

	R(a, b, c, d, e, F1, K1, x[ 0]);
	R(e, a, b, c, d, F1, K1, x[ 1]);
	R(d, e, a, b, c, F1, K1, x[ 2]);
	R(c, d, e, a, b, F1, K1, x[ 3]);
	R(b, c, d, e, a, F1, K1, x[ 4]);
	R(a, b, c, d, e, F1, K1, x[ 5]);
	R(e, a, b, c, d, F1, K1, x[ 6]);
	R(d, e, a, b, c, F1, K1, x[ 7]);
	R(c, d, e, a, b, F1, K1, x[ 8]);
	R(b, c, d, e, a, F1, K1, x[ 9]);
	R(a, b, c, d, e, F1, K1, x[10]);
	R(e, a, b, c, d, F1, K1, x[11]);
	R(d, e, a, b, c, F1, K1, x[12]);
	R(c, d, e, a, b, F1, K1, x[13]);
	R(b, c, d, e, a, F1, K1, x[14]);
	R(a, b, c, d, e, F1, K1, x[15]);
	R(e, a, b, c, d, F1, K1, M(16));
	R(d, e, a, b, c, F1, K1, M(17));
	R(c, d, e, a, b, F1, K1, M(18));
	R(b, c, d, e, a, F1, K1, M(19));
	R(a, b, c, d, e, F2, K2, M(20));
	R(e, a, b, c, d, F2, K2, M(21));
	R(d, e, a, b, c, F2, K2, M(22));
	R(c, d, e, a, b, F2, K2, M(23));
	R(b, c, d, e, a, F2, K2, M(24));
	R(a, b, c, d, e, F2, K2, M(25));
	R(e, a, b, c, d, F2, K2, M(26));
	R(d, e, a, b, c, F2, K2, M(27));
	R(c, d, e, a, b, F2, K2, M(28));
	R(b, c, d, e, a, F2, K2, M(29));
	R(a, b, c, d, e, F2, K2, M(30));
	R(e, a, b, c, d, F2, K2, M(31));
	R(d, e, a, b, c, F2, K2, M(32));
	R(c, d, e, a, b, F2, K2, M(33));
	R(b, c, d, e, a, F2, K2, M(34));
	R(a, b, c, d, e, F2, K2, M(35));
	R(e, a, b, c, d, F2, K2, M(36));
	R(d, e, a, b, c, F2, K2, M(37));
	R(c, d, e, a, b, F2, K2, M(38));
	R(b, c, d, e, a, F2, K2, M(39));
	R(a, b, c, d, e, F3, K3, M(40));
	R(e, a, b, c, d, F3, K3, M(41));
	R(d, e, a, b, c, F3, K3, M(42));
	R(c, d, e, a, b, F3, K3, M(43));
	R(b, c, d, e, a, F3, K3, M(44));
	R(a, b, c, d, e, F3, K3, M(45));
	R(e, a, b, c, d, F3, K3, M(46));
	R(d, e, a, b, c, F3, K3, M(47));
	R(c, d, e, a, b, F3, K3, M(48));
	R(b, c, d, e, a, F3, K3, M(49));
	R(a, b, c, d, e, F3, K3, M(50));
	R(e, a, b, c, d, F3, K3, M(51));
	R(d, e, a, b, c, F3, K3, M(52));
	R(c, d, e, a, b, F3, K3, M(53));
	R(b, c, d, e, a, F3, K3, M(54));
	R(a, b, c, d, e, F3, K3, M(55));
	R(e, a, b, c, d, F3, K3, M(56));
	R(d, e, a, b, c, F3, K3, M(57));
	R(c, d, e, a, b, F3, K3, M(58));
	R(b, c, d, e, a, F3, K3, M(59));
	R(a, b, c, d, e, F4, K4, M(60));
	R(e, a, b, c, d, F4, K4, M(61));
	R(d, e, a, b, c, F4, K4, M(62));
	R(c, d, e, a, b, F4, K4, M(63));
	R(b, c, d, e, a, F4, K4, M(64));
	R(a, b, c, d, e, F4, K4, M(65));
	R(e, a, b, c, d, F4, K4, M(66));
	R(d, e, a, b, c, F4, K4, M(67));
	R(c, d, e, a, b, F4, K4, M(68));
	R(b, c, d, e, a, F4, K4, M(69));
	R(a, b, c, d, e, F4, K4, M(70));
	R(e, a, b, c, d, F4, K4, M(71));
	R(d, e, a, b, c, F4, K4, M(72));
	R(c, d, e, a, b, F4, K4, M(73));
	R(b, c, d, e, a, F4, K4, M(74));
	R(a, b, c, d, e, F4, K4, M(75));
	R(e, a, b, c, d, F4, K4, M(76));
	R(d, e, a, b, c, F4, K4, M(77));
	R(c, d, e, a, b, F4, K4, M(78));
	R(b, c, d, e, a, F4, K4, M(79));

	/* Update chaining vars. */
	hd->h0 += a;
	hd->h1 += b;
	hd->h2 += c;
	hd->h3 += d;
	hd->h4 += e;
}


/*
 * Update the message digest with the contents of INBUF with length INLEN.
 */
static void sha1_write(SHA1_CONTEXT *hd, const unsigned char *inbuf,
    size_t inlen)
{
	if (hd->count == 64) {
		/* Flush the buffer. */
		transform( hd, hd->buf );
		burn_stack(88 + 4 * sizeof(void *));
		hd->count = 0;
		hd->nblocks++;
	}
	
	if (!inbuf)
		return;
	
	if (hd->count) {
		for( ; inlen && hd->count < 64; inlen--)
			hd->buf[hd->count++] = *inbuf++;

		sha1_write(hd, NULL, 0);

		if (!inlen)
			return;
	}

	while (inlen >= 64) {
		transform(hd, inbuf);
		hd->count = 0;
		hd->nblocks++;
		inlen -= 64;
		inbuf += 64;
	}
	
	burn_stack(88 + 4 * sizeof(void *));

	for( ; inlen && hd->count < 64; inlen--)
		hd->buf[hd->count++] = *inbuf++;
}


/*
 * The routine final terminates the computation and returns the digest.  The
 * handle is prepared for a new cycle, but adding bytes to the handle will the
 * destroy the returned buffer.
 *
 * Returns: 20 bytes representing the digest.
 */

static void sha1_final(SHA1_CONTEXT *hd)
{
	uint32 t, msb, lsb;
	unsigned char *p;

	/* Flush. */
	sha1_write(hd, NULL, 0);

	t = hd->nblocks;
	
	/* Multiply by 64 to make a byte count. */
	lsb = t << 6;
	msb = t >> 26;

	/* Add the count. */
	t = lsb;
	if ((lsb += hd->count) < t)
		msb++;

	/* Multiply by 8 to make a bit count. */
	t = lsb;
	lsb <<= 3;
	msb <<= 3;
	msb |= t >> 29;

	if (hd->count < 56) {
		/* Enough room. */
		/* Pad. */
		hd->buf[hd->count++] = 0x80;

		while (hd->count < 56) {
			/* Pad. */
			hd->buf[hd->count++] = 0;
		}
	} else {
		/* Need one extra block. */
		/* Pad. */
		hd->buf[hd->count++] = 0x80;

		while (hd->count < 64)
			hd->buf[hd->count++] = 0;

		/* Flush. */		
		sha1_write(hd, NULL, 0);

		/* Fill next block with zeroes. */
		memset(hd->buf, 0, 56);
	}

	/* Append the 64 bit count. */
	hd->buf[56] = msb >> 24;
	hd->buf[57] = msb >> 16;
	hd->buf[58] = msb >>  8;
	hd->buf[59] = msb      ;
	hd->buf[60] = lsb >> 24;
	hd->buf[61] = lsb >> 16;
	hd->buf[62] = lsb >>  8;
	hd->buf[63] = lsb      ;

	transform(hd, hd->buf);
	burn_stack(88 + 4 * sizeof(void *));

	p = hd->buf;

#ifdef WORDS_BIGENDIAN
#	define X(a) do { *(uint32*)p = hd->h##a ; p += 4; } while(0)
#else
	/* Little endian. */
#	define X(a) do { *p++ = hd->h##a >> 24; *p++ = hd->h##a >> 16;	 \
		      *p++ = hd->h##a >> 8; *p++ = hd->h##a; } while(0)
#endif

	X(0);
	X(1);
	X(2);
	X(3);
	X(4);
#undef X
}

static unsigned char *sha1_read(SHA1_CONTEXT *hd)
{
	return(hd->buf);
}

/*
 * Return some information about the algorithm.  We need algo here to
 * distinguish different flavors of the algorithm.
 *
 * Returns: A pointer to string describing the algorithm or NULL if the ALGO is
 *          invalid.
 */
const char *sha1_get_info(int algo, size_t *contextsize,
    unsigned char **r_asnoid, int *r_asnlen, int *r_mdlen,
    void (**r_init)(void *c),
    void (**r_write)(void *c, unsigned char *buf, size_t nbytes),
    void (**r_final)(void *c), unsigned char *(**r_read)(void *c))
{
	/* Object ID is 1.3.14.3.2.26 */
	static unsigned char asn[15] = {
	    0x30, 0x21, 0x30, 0x09, 0x06, 0x05, 0x2b, 0x0e, 0x03, 0x02,
	    0x1a, 0x05, 0x00, 0x04, 0x14
	};
	
	if (algo != 2)
		return NULL;

	*contextsize = sizeof(SHA1_CONTEXT);
	*r_asnoid = asn;
	*r_asnlen = lenof(asn);
	*r_mdlen = 20;
	*(void (**)(SHA1_CONTEXT *))r_init                                 = sha1_init;
	*(void (**)(SHA1_CONTEXT *, const unsigned char *, size_t))r_write = sha1_write;
	*(void (**)(SHA1_CONTEXT *))r_final                                = sha1_final;
	*(unsigned char *(**)(SHA1_CONTEXT *))r_read                       = sha1_read;

	return("SHA1");
}

static char bin2hex[] = {
	'0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'a', 'b', 'c',
	'd', 'e', 'f'
};

/*
 * SHA-1 returns 160 bits of digest (20 bytes), which we turn into 40 hex
 * characters and a terminating \0.  Therefore buffer must be a preallocated
 * buffer of at least 41 bytes.
 */ 
void make_sha_hash(const char *passwd, char *buffer)
{
	SHA1_CONTEXT ctx;
	int i, j, high, low;
	unsigned char *shadigest;

	sha1_init(&ctx);
	sha1_write(&ctx, passwd, strlen(passwd));
	sha1_final(&ctx);
	shadigest = sha1_read(&ctx);

        for (i = 0, j = 0; i < 20; i++, j += 2) {
                high = shadigest[i] >> 4;
                low = shadigest[i] - (high << 4);
                buffer[j] = bin2hex[high];
                buffer[j + 1] = bin2hex[low];
        }

	buffer[40] = 0;
}

